### DISCUSSION TITLE
NodeMCU ESP8266 module connections

### DISCUSSION THREAD

**Discussion by jlt206** [Bodmer/TFT_eSPI#2482] [Q&A - General]
Hi, I am working with a AZ NodeMCU ESP8266 module and a Pimoroni 1.3" 240x240 Round LCD but I'm struggling with matching the SPI interfaces? I am using D1, D2 and A0 for other sensors already.

LCD
3.5v - 3.5v
CS -
SCK -
MOSI -
DC -
BL - Backlight connected to 3.5v
GND - GND

**Comment by Bodmer** [comment]
Example connections here:
https://github.com/Bodmer/TFT_eSPI/blob/master/User_Setup.h#L137

**Reply by jlt206** [reply]
Thank you. Sorry for the wrong location I am new to working with Github.

I wil try these settings.

Cheers!

**Reply by jlt206** [reply]
So.. I tried another Library and had similar issues. Then by accident I got a good result by using W-240 and H-280? but have set it ti 260 x 260 in the Setup File for now.
The specification 240x240 pixels (~260 PPI)?
https://shop.pimoroni.com/products/1-3-spi-colour-round-lcd-240x240-breakout

I also get errors for this code using the watchESP
    ledcSetup(pwmLedChannelTFT, pwmFreq, pwmResolution);
    ledcAttachPin(5, pwmLedChannelTFT);
    ledcWrite(pwmLedChannelTFT, 200);

**Comment by jlt206** [comment]
![Screen Test](https://user-images.githubusercontent.com/4954838/224041852-1106fc8a-3d0a-45be-826b-1cbe4db90101.png)

I've been able to get it working, but I can't seem to get the tests to fill the screen correctly, not sure if this is hardware, or config?

But the fill images seem to be too high, so it leaves the bottom of the screen not covered as per the image.

**Reply by Bodmer** [reply]
The display controller is for a 240 x 320 display so you can get an 80 pixel offset.

Set the width and height to 240.

Use tft.setRotation(1) and you may find the offset is then zero.

**Reply by jlt206** [reply]
![PXL_20230310_084124905 MP](https://user-images.githubusercontent.com/4954838/224268320-4050f065-9e3d-4e80-b737-0eb8e7efe673.jpg)

At the moment I have had to use the Adafruit librbary but today will try yours again. 

I did rotate the screen to the correct oreintation but still have to use 240x280 to get the full coverage. 

Is there way to add an offset?

**Reply by Bodmer** [reply]
Yes, the following lines should be used in the setup file:
```
#define ST7789_DRIVER     // Configure all registers

#define TFT_WIDTH  240
#define TFT_HEIGHT 240

#define CGRAM_OFFSET      // Library will add offsets required

```

**Comment by jlt206** [comment]
![PXL_20230310_122829679](https://user-images.githubusercontent.com/4954838/224316602-01a198d1-4d8d-4ff8-8de9-29893cd39c9a.jpg)

So I made the changes as suggested and wrote a test as below.

#include "SPI.h"
#include "TFT_eSPI.h"

TFT_eSPI tft = TFT_eSPI();

unsigned long total = 0;
unsigned long tn = 0;
void setup() {
  Serial.begin(115200);

  while (!Serial);

  Serial.println(""); Serial.println("");
  Serial.println("TFT_eSPI library test!");

  tft.init();
  tft.setRotation(0);
  tft.fillScreen(TFT_BLUE);
  tft.fillCircle(120,120, 120, TFT_BLACK);
  tft.fillCircle(120,120, 20, TFT_RED);
//  tft.fillCircle(118, 118, 4, TFT_YELLOW);
  tft.setTextSize(4);
  tft.setTextColor(TFT_WHITE);
  tft.setCursor(117,117);
  tft.println("A");
}

The only way I can get the screen to fill so far is with setting 240x280.

The imge shows the results I have. So the offset does not seem to be applied?

**Comment by Bodmer** [comment]
Your test results indicate the y offset for that display variant is 40, so that is not compensated for in the library at the moment. 

Comment out these lines in your setup file:
```
//#define TFT_WIDTH  240
//#define TFT_HEIGHT 320
//#define CGRAM_OFFSET      // Library will add offsets required
```
This will set the screen area to 240x320.

The use setViewport to compensate for the offset in the sketch thus:
```
tft.init();
tft.setRotation(0);
tft.setViewport(0, 40, 240, 240);
```

This will move the display area and then the display area origin should be corrected.

**Comment by jlt206** [comment]
![PXL_20230311_090744681](https://user-images.githubusercontent.com/4954838/224475750-002f9229-1cd5-4ca2-9031-fdbbb2a5e56a.jpg)

Thanks. That seems to have worked.
I have noticed that there is a problem with selecting colors though. The TFT_RED and TFT_BLUE seem to be reversed and I couldn't select TFT_YELLOW.

Where can these be eddited and colors added so I can add custom colors please?

#include "SPI.h"
#include "TFT_eSPI.h"

TFT_eSPI tft = TFT_eSPI();

unsigned long total = 0;
unsigned long tn = 0;
void setup() {
  Serial.begin(115200);

  while (!Serial);

  Serial.println(""); Serial.println("");
  Serial.println("TFT_eSPI library test!");

  tft.init();
  tft.setRotation(0);
  tft.setViewport(0, 40, 240, 240);

  tft.fillScreen(TFT_BLACK);
  delay(2000);
  tft.fillCircle(120,120,120, TFT_RED);
  delay(2000);
  tft.fillCircle(120,120,80, TFT_WHITE);
  delay(2000);
  tft.fillCircle(120,120,20, TFT_BLUE);
//  tft.fillCircle(118, 118, 4, TFT_YELLOW);
  tft.setTextSize(4);
  tft.setTextColor(TFT_WHITE);
  tft.setCursor(110,106);
  tft.println("A");
}

**Reply by Bodmer** [reply]
Add ONE of these lines to the setup file:
#define TFT_RGB_ORDER TFT_RGB  // Colour order Red-Green-Blue
#define TFT_RGB_ORDER TFT_BGR  // Colour order Blue-Green-Red

Try each one to see which corrects the colours. Let me know which one worked.
